### PR TITLE
Php scl

### PR DESCRIPTION
**Pull request by stephdl**
Install the new virtualhost rpm
create a new virtualhost

set the PhpRhVersion to `default` or `php71` or `php72`

The relevant pool of php scl will be installed by the UI, php settings are adjustable per Vhost, no matter what php version is chosen.

https://github.com/NethServer/dev/issues/5912

----
OBSOLETE: 
`yum install nethserver-rh-php72-php-fpm nethserver-rh-php71-php-fpm -y`
```
db vhosts setprop db2d5cb47968232 PhpRhVersion php71
signal-event vhost-modify 
```

you can check the pool is running

```
systemctl status rh-php71-php-fpm
less /etc/opt/rh/rh-php71/php-fpm.d/php-fpm.db2d5cb47968232.conf
ll /var/run/rh-php71-php-fpm/db2d5cb47968232-php71.sock
```

### CODE REVIEW

File: api/virtualhost/update [hunk 1]
```diff
@@ -64,6 +64,12 @@ if ($vhostName eq 'default'){
         'ServerNames' => lc(join (',',uniq (@{$input->{'virtualhost'}->{'ServerNames'}}))),
         'SslCertificate' => $input->{'virtualhost'}->{'SslCertificate'},
         'status' => $input->{'virtualhost'}->{'status'},
+        "MaxExecutionTime" => $input->{'virtualhost'}->{'MaxExecutionTime'},
+        "MemoryLimit" => $input->{'virtualhost'}->{'MemoryLimit'},
+        "PostMaxSize" => $input->{'virtualhost'}->{'PostMaxSize'},
+        "UploadMaxFilesize" => $input->{'virtualhost'}->{'UploadMaxFilesize'},
+        "PhpRhVersion" => $input->{'virtualhost'}->{'PhpRhVersion'},
+        "phpCustomSettings" => $input->{'virtualhost'}->{'phpCustomSettings'}
```
Comment (DavidePrincipi) [inline api/virtualhost/update]: Is there any reason to use lowercase here?

```suggestion
        "PhpCustomSettings" => $input->{'virtualhost'}->{'phpCustomSettings'}
```

File: nethserver-httpd.spec [hunk 2]
```diff
@@ -90,6 +90,9 @@ done
 
 
 %changelog
+* Mon Oct 28 2019 Giacomo Sanchietti <giacomo.sanchietti@nethesis.it> - 3.6.2-1
```
Comment (DavidePrincipi) [inline nethserver-httpd.spec]: I guess there was a merge-conflict  here, right?

File: ui/src/components/ModalVhostEdit.vue [hunk 3]
```diff
@@ -352,6 +352,155 @@ select {
...
+                    <label class="col-sm-3 control-label">{{$t('virtualhost.MaxExecutionTime')}}
+                        <doc-info
+                          :placement="'top'"
+                          :title="$t('virtualhost.MaxExecutionTime')"
+                          :chapter="'MaxExecutionTime'"
+                          :inline="true"
+                        ></doc-info>
+                    </label>
+                    <div class="col-sm-9">
+                        <div v-if="MaxExecutionTime > 0">{{MaxExecutionTime + ' '+ $t('virtualhost.seconds')}}</div>
```
Comment (DavidePrincipi) [inline ui/src/components/ModalVhostEdit.vue]: The vue-i18n extension has support for both formatting and pluralization

- https://kazupon.github.io/vue-i18n/guide/formatting.html
- https://kazupon.github.io/vue-i18n/guide/pluralization.html

For instance 
- (formatting) https://github.com/NethServer/nethserver-samba/blob/6bc69adc70f070ebb2e0c8379bfb1b5aba42b3dd/ui/src/components/SharedFolderEditModal.vue#L78
- (pluralization) - can't find it

File: ui/src/components/VhostsListView.vue [hunk 4]
```diff
@@ -95,6 +95,9 @@
             <div class="list-group-item-text">
               <div :class="[item.status === 'disabled' ? 'gray':'']">{{ item.Description }}</div>
             </div>
+            <div class="list-group-item-text">
+              <div :class="[item.status === 'disabled' ? 'gray':'']">{{(item.PhpRhVersion ==='default')? 'PHP54' : item.PhpRhVersion.toUpperCase()}}</div>
```
Comment (DavidePrincipi) [inline ui/src/components/VhostsListView.vue]: Can we use a more "traditional" string, with spaces and periods, like `PHP 5.4`? The `$t()` translation function might help...

```suggestion
              <div :class="[item.status === 'disabled' ? 'gray':'']">{{ $t('virtualhost.php_version_' + item.PhpRhVersion) }}</div>
```

File: virtualhosts/etc/e-smith/templates/etc/opt/rh/rh-php71/php-fpm.d/000-virtualhost.conf/10Base [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+pm.start_servers = 5
+pm.min_spare_servers = 5
+pm.max_spare_servers = 35
+
+user = apache;
+group = apache
+listen.owner = root
+listen.group = apache
+listen.mode = 0660
+listen.allowed_clients = 127.0.0.1
```
Comment (DavidePrincipi) [inline virtualhosts/etc/e-smith/templates/etc/opt/rh/rh-php71/php-fpm.d/000-virtualhost.conf/10Base]: Is it still needed for Unix sockets?

```suggestion
```

File: virtualhosts/etc/e-smith/templates/etc/opt/rh/rh-php72/php-fpm.d/000-virtualhost.conf/10Base [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+pm.start_servers = 5
+pm.min_spare_servers = 5
+pm.max_spare_servers = 35
+
+user = apache;
+group = apache
+listen.owner = root
+listen.group = apache
+listen.mode = 0660
+listen.allowed_clients = 127.0.0.1
```
Comment (DavidePrincipi) [inline virtualhosts/etc/e-smith/templates/etc/opt/rh/rh-php72/php-fpm.d/000-virtualhost.conf/10Base]: Same as above...
```suggestion
```

File: virtualhosts/etc/e-smith/templates/etc/opt/rh/rh-php71/php-fpm.d/000-virtualhost.conf/10Base [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        $OUT .= qq(
+;
+; vhosts $ServerNames
+;
+
+[$VhostName]
+listen = /var/run/rh-$PhpRhVersion-php-fpm/$VhostName-$PhpRhVersion.sock
+
+;Logs
+php_admin_value[error_log] = /var/opt/rh/rh-$PhpRhVersion/log/php-fpm/error.log
```
Comment (DavidePrincipi) [inline virtualhosts/etc/e-smith/templates/etc/opt/rh/rh-php71/php-fpm.d/000-virtualhost.conf/10Base]: Do we have logrotate configuration?

File: virtualhosts/etc/e-smith/templates/etc/opt/rh/rh-php72/php-fpm.d/000-virtualhost.conf/10Base [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+        $OUT .= qq(
+;
+; vhosts $ServerNames
+;
+
+[$VhostName]
+listen = /var/run/rh-$PhpRhVersion-php-fpm/$VhostName-$PhpRhVersion.sock
+
+;Logs
+php_admin_value[error_log] = /var/opt/rh/rh-$PhpRhVersion/log/php-fpm/error.log
```
Comment (DavidePrincipi) [inline virtualhosts/etc/e-smith/templates/etc/opt/rh/rh-php72/php-fpm.d/000-virtualhost.conf/10Base]: Do we have logrotate configuration?

**Review by DavidePrincipi — COMMENTED** [review]
This is a great job! I didn't check out the latest RPM, however from the code POV it seems almost complete. Please consider the following code tweaks - I cannot promise to not add more ;)

File: nethserver-httpd.spec (hunk 2 above)
Comment (stephdl) [reply nethserver-httpd.spec]: tk it should be solved now

File: virtualhosts/etc/e-smith/templates/etc/opt/rh/rh-php72/php-fpm.d/000-virtualhost.conf/10Base (hunk 8 above)
Comment (stephdl) [reply virtualhosts/etc/e-smith/templates/etc/opt/rh/rh-php72/php-fpm.d/000-virtualhost.conf/10Base]: yes from upstream rpm

```
/var/opt/rh/rh-php72/log/php-fpm/*log {
    missingok
    notifempty
    sharedscripts
    delaycompress
    postrotate
        /bin/kill -SIGUSR1 `cat /var/opt/rh/rh-php72/run/php-fpm/php-fpm.pid 2>/dev/null` 2>/dev/null || true
    endscript
}
```

File: virtualhosts/etc/e-smith/templates/etc/opt/rh/rh-php71/php-fpm.d/000-virtualhost.conf/10Base (hunk 7 above)
Comment (stephdl) [reply virtualhosts/etc/e-smith/templates/etc/opt/rh/rh-php71/php-fpm.d/000-virtualhost.conf/10Base]: yes from upstream rpm

```
/var/opt/rh/rh-php71/log/php-fpm/*log {
    missingok
    notifempty
    sharedscripts
    delaycompress
    postrotate
        /bin/kill -SIGUSR1 `cat /var/opt/rh/rh-php71/run/php-fpm/php-fpm.pid 2>/dev/null` 2>/dev/null || true
    endscript
}
```

File: virtualhosts/etc/e-smith/templates/etc/opt/rh/rh-php72/php-fpm.d/000-virtualhost.conf/10Base (hunk 6 above)
Comment (stephdl) [reply virtualhosts/etc/e-smith/templates/etc/opt/rh/rh-php72/php-fpm.d/000-virtualhost.conf/10Base]: removed

File: virtualhosts/etc/e-smith/templates/etc/opt/rh/rh-php71/php-fpm.d/000-virtualhost.conf/10Base (hunk 5 above)
Comment (stephdl) [reply virtualhosts/etc/e-smith/templates/etc/opt/rh/rh-php71/php-fpm.d/000-virtualhost.conf/10Base]: removed

File: api/virtualhost/update (hunk 1 above)
Comment (stephdl) [reply api/virtualhost/update]: done

File: ui/src/components/VhostsListView.vue (hunk 4 above)
Comment (stephdl) [reply ui/src/components/VhostsListView.vue]: done

File: ui/src/components/ModalVhostEdit.vue (hunk 3 above)
Comment (stephdl) [reply ui/src/components/ModalVhostEdit.vue]: pluralization done

File: virtualhosts/etc/e-smith/templates/etc/opt/rh/rh-php72/php-fpm.d/000-virtualhost.conf/10Base [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
```
Comment (DavidePrincipi) [inline virtualhosts/etc/e-smith/templates/etc/opt/rh/rh-php72/php-fpm.d/000-virtualhost.conf/10Base]: Is it reasonable to split error log by pool? Is it safe to use `$VhostName` in the error log file name (i.e. old nethgui vhosts have a name set by the user)

```suggestion
php_admin_value[error_log] = /var/opt/rh/rh-$PhpRhVersion/log/php-fpm/error-$VhostName.log
```

**Review by DavidePrincipi — COMMENTED** [review]
What about splitting the error logs by pool name?

File: virtualhosts/etc/e-smith/templates/etc/opt/rh/rh-php72/php-fpm.d/000-virtualhost.conf/10Base [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+            my $MaxFileUploads = $vdb->get_prop("$VhostName",'MaxFileUploads') || 'disabled';
+            
+            $OUT .= "php_admin_flag[allow_url_fopen] = off\n" if ($AllowUrlfOpen eq 'disabled');
+            $OUT .= "php_admin_value[memory_limit] = $MemoryLimit"."M\n" if ($MemoryLimit ne '0');
+            $OUT .= "php_admin_value[upload_max_filesize] = $UpMaxFileSize"."M\n" if ($UpMaxFileSize ne '0');
+            $OUT .= "php_admin_value[post_max_size] = $PostMaxSize"."M\n" if ($PostMaxSize ne '0');
+            $OUT .= "php_admin_value[max_execution_time] = $MaxExecTime\n" if ($MaxExecTime ne 'disabled');
+            $OUT .= "php_admin_value[max_input_time] =  $MaxExecTime\n" if ($MaxExecTime ne 'disabled');
+            $OUT .= "php_admin_value[max_file_uploads] = $MaxFileUploads\n" if ($MaxFileUploads ne 'disabled');
+        }
```
Comment (DavidePrincipi) [inline virtualhosts/etc/e-smith/templates/etc/opt/rh/rh-php72/php-fpm.d/000-virtualhost.conf/10Base:86]: Why not adding here an `else` branch?

If custom settings are disabled we set the defaults values from the `php` key settings in ConfigDB:

```text
[root@vm5 test]# config show php
php=configuration
    DateTimezone=UTC
    ExposePhp=
    MaxExecutionTime=0
    MemoryLimit=256
    PostMaxSize=8
    ShortOpenTag=
    UploadMaxFilesize=2
```

However there are some hardcoded default settings to take into account...

https://github.com/NethServer/nethserver-httpd/blob/96770d692d0213d5e7099481ed4a0f390bb1bd51/ui/src/views/Settings.vue#L107-L112

**Review by DavidePrincipi — COMMENTED** [review]
Inherit pool default values from PHP settings

File: virtualhosts/etc/e-smith/templates/etc/opt/rh/rh-php72/php-fpm.d/000-virtualhost.conf/10Base (hunk 9 above)
Comment (stephdl) [reply virtualhosts/etc/e-smith/templates/etc/opt/rh/rh-php72/php-fpm.d/000-virtualhost.conf/10Base]: this is what I did first, but I think it is easier like apache to get all error in one file rather to looks after all, but I can reverse...sometime yo prefer to get all in one file (like configuration for example)

Comment (DavidePrincipi) [reply virtualhosts/etc/e-smith/templates/etc/opt/rh/rh-php72/php-fpm.d/000-virtualhost.conf/10Base]: I guess there isn't any reference to the pool name. If I have multiple applications and virtual hosts separate log files can help to understand the message/issue origin.

Comment (stephdl) [reply virtualhosts/etc/e-smith/templates/etc/opt/rh/rh-php72/php-fpm.d/000-virtualhost.conf/10Base]: ok :D

Comment (stephdl) [reply virtualhosts/etc/e-smith/templates/etc/opt/rh/rh-php72/php-fpm.d/000-virtualhost.conf/10Base]: done

File: virtualhosts/etc/e-smith/templates/etc/opt/rh/rh-php72/php-fpm.d/000-virtualhost.conf/10Base (hunk 10 above)
Comment (stephdl) [reply virtualhosts/etc/e-smith/templates/etc/opt/rh/rh-php72/php-fpm.d/000-virtualhost.conf/10Base:86]: not sure to understand well, if PhpCustomSettings is disabled, then the values comes from the php.ini of the php pool IIRC, why do you want to enable others values

Comment (stephdl) [reply virtualhosts/etc/e-smith/templates/etc/opt/rh/rh-php72/php-fpm.d/000-virtualhost.conf/10Base:86]: yep confirmed, the values come from `/etc/opt/rh/rh-php72/php.ini` if not set by our UI default are

```
max_execution_time	30
memory_limit	128M
post_max_size	8M
upload_max_filesize	2M
```

do you prefer others values from our php settings (only differences is `MaxExecutionTime`)

Comment (DavidePrincipi) [reply virtualhosts/etc/e-smith/templates/etc/opt/rh/rh-php72/php-fpm.d/000-virtualhost.conf/10Base:86]: Yes, if custom values are not enabled I'd expect values from the Settings page are applied. There is no other information provided by the UI about the PHP configurations.

Comment (stephdl) [reply virtualhosts/etc/e-smith/templates/etc/opt/rh/rh-php72/php-fpm.d/000-virtualhost.conf/10Base:86]: ok, lets code

File: virtualhosts/etc/e-smith/templates/etc/opt/rh/rh-php71/php-fpm.d/000-virtualhost.conf/10Base [hunk 11]
```diff
@@ -68,6 +69,18 @@ php_value[soap.wsdl_cache_dir]  = /var/opt/rh/rh-$PhpRhVersion/lib/php/wsdlcache
...
+            my $MemoryLimit = $db->get_prop('php','MemoryLimit') || '128';
+            my $UpMaxFileSize = $db->get_prop('php','UploadMaxFilesize') || '2';
+            my $PostMaxSize = $db->get_prop('php','PostMaxSize') || '8';
+            my $MaxExecTime = $db->get_prop('php','MaxExecutionTime') || '0';
+            
+            $OUT .= "php_admin_value[memory_limit] = $MemoryLimit"."M\n" if ($MemoryLimit ne '0');
+            $OUT .= "php_admin_value[upload_max_filesize] = $UpMaxFileSize"."M\n" if ($UpMaxFileSize ne '0');
+            $OUT .= "php_admin_value[post_max_size] = $PostMaxSize"."M\n" if ($PostMaxSize ne '0');
+            $OUT .= "php_admin_value[max_execution_time] = $MaxExecTime\n" if ($MaxExecTime ne 'disabled');
+            $OUT .= "php_admin_value[max_input_time] =  $MaxExecTime\n" if ($MaxExecTime ne 'disabled');
```
Comment (DavidePrincipi) [inline virtualhosts/etc/e-smith/templates/etc/opt/rh/rh-php71/php-fpm.d/000-virtualhost.conf/10Base]: I feel the `...if ()` can be removed. Isn't the zero (`0`) value implicitly the definition of "disabled"? What are the differences?

Same question for existing lines 65-71: why `...if()` is needed?

**Review by DavidePrincipi — COMMENTED** [review]
Question about `...if()`s

File: ui/public/i18n/language.json [hunk 12]
```diff
@@ -95,7 +95,28 @@
     "dnsHostRecord_do_not_be_removed": "Host name aliases in System > Dashboard > Hostname are preserved",
     "http": "HTTP auth",
     "ftp": "FTP access",
-    "vsftpd_is_not_running": "the FTP service is not running"
+    "vsftpd_is_not_running": "the FTP service is not running",
+    "default_php_version": "Default",
```
Comment (DavidePrincipi) [inline ui/public/i18n/language.json]:
```suggestion
    "default_php_version": "PHP 5.4 (default)",
```

**Review by DavidePrincipi — COMMENTED** [review]
Just a little label change proposal...

File: ui/public/i18n/language.json (hunk 12 above)
Comment (stephdl) [reply ui/public/i18n/language.json]: done

File: virtualhosts/etc/e-smith/templates/etc/opt/rh/rh-php71/php-fpm.d/000-virtualhost.conf/10Base (hunk 11 above)
Comment (stephdl) [reply virtualhosts/etc/e-smith/templates/etc/opt/rh/rh-php71/php-fpm.d/000-virtualhost.conf/10Base]: done, need to test now

Comment (stephdl) [reply virtualhosts/etc/e-smith/templates/etc/opt/rh/rh-php71/php-fpm.d/000-virtualhost.conf/10Base]: good, resolved

File: virtualhosts/etc/e-smith/templates/etc/opt/rh/rh-php71/php-fpm.d/000-virtualhost.conf/10Base [hunk 13]
```diff
@@ -56,31 +56,31 @@ php_value[soap.wsdl_cache_dir]  = /var/opt/rh/rh-$PhpRhVersion/lib/php/wsdlcache
...
-            $OUT .= "php_admin_value[upload_max_filesize] = $UpMaxFileSize"."M\n" if ($UpMaxFileSize ne '0');
-            $OUT .= "php_admin_value[post_max_size] = $PostMaxSize"."M\n" if ($PostMaxSize ne '0');
-            $OUT .= "php_admin_value[max_execution_time] = $MaxExecTime\n" if ($MaxExecTime ne 'disabled');
-            $OUT .= "php_admin_value[max_input_time] =  $MaxExecTime\n" if ($MaxExecTime ne 'disabled');
+            $OUT .= "php_admin_value[memory_limit] = $MemoryLimit"."M\n";
+            $OUT .= "php_admin_value[upload_max_filesize] = $UpMaxFileSize"."M\n";
+            $OUT .= "php_admin_value[post_max_size] = $PostMaxSize"."M\n";
+            $OUT .= "php_admin_value[max_execution_time] = $MaxExecTime\n";
+            $OUT .= "php_admin_value[max_input_time] =  $MaxExecTime\n";
             $OUT .= "php_admin_value[max_file_uploads] = $MaxFileUploads\n" if ($MaxFileUploads ne 'disabled');
```
Comment (DavidePrincipi) [inline virtualhosts/etc/e-smith/templates/etc/opt/rh/rh-php71/php-fpm.d/000-virtualhost.conf/10Base]: Can we handle `max_file_uploads` in a way consistent with other settings (i.e. fall back to ConfigDB `php` props)?

It seems the PHP default for max_file_uploads is `20`:

https://www.php.net/manual/en/ini.core.php#ini.sect.file-uploads

**Review by DavidePrincipi — COMMENTED** [review]
Also `max_file_uploads`?

File: virtualhosts/etc/e-smith/templates/etc/opt/rh/rh-php71/php-fpm.d/000-virtualhost.conf/10Base (hunk 13 above)
Comment (stephdl) [reply virtualhosts/etc/e-smith/templates/etc/opt/rh/rh-php71/php-fpm.d/000-virtualhost.conf/10Base]:
```
[root@ns7loc11 ~]# db configuration show php 
php=configuration
    DateTimezone=Europe/Paris
    ExposePhp=
    MaxExecutionTime=2006
    MemoryLimit=280
    PostMaxSize=152
    ShortOpenTag=
    UploadMaxFilesize=76
```
hum we do not have it inside our prop ?

Comment (DavidePrincipi) [reply virtualhosts/etc/e-smith/templates/etc/opt/rh/rh-php71/php-fpm.d/000-virtualhost.conf/10Base]: Let's add on more ;)

**Review by DavidePrincipi — APPROVED** [review]
Let's go to test this. 

:warning: Please add also the new props documentation for virtualhosts e-smith DB, and show how they fall back to configuration/php props as defaults.

**Pull request merged**